File: 백준/Gold/1717. 집합의 표현/집합의 표현.cc
#include <iostream>
#include <set>
using namespace std;
int n, m;
int parent[1000002];
int get_parent(int parent[], int x) {
	if (parent[x] == x) return x;
	return parent[x] = get_parent(parent, parent[x]);
}
void union_parent(int parent[], int a, int b) {
	a = get_parent(parent, a);
	b = get_parent(parent, b);
	if (a < b)		parent[b] = a;
	else if (a > b) parent[a] = b;
	else			return;
}

int main() {
	ios_base::sync_with_stdio(0);
	cin.tie(0); cout.tie(0);
	cin >> n >> m;
	for (int i = 0; i < n + 1; i++)
		parent[i] = i;
	for (int i = 1; i < m + 1; i++) {
		bool b;
		int v1, v2;
		cin >> b >>v1 >> v2;
		if (b) {
			if (get_parent(parent, v1) == get_parent(parent, v2))
				cout << "YES\n";
			else
				cout << "NO\n";
		}
		else
			union_parent(parent, v1, v2);
	}	
	return 0;
}

File: 백준/Gold/1717. 집합의 표현/집합의 표현.py
import sys
sys.setrecursionlimit(10 ** 9)


def get_parent(p, x):
    if p[x] != x:
        p[x] = get_parent(p, p[x])
    return p[x]


def union_parent(p, a, b):
    a = get_parent(p, a)
    b = get_parent(p, b)
    if a < b:
        p[b] = a
    elif a > b:
        p[a] = b


def solution():
    v, e = map(int, sys.stdin.readline().rstrip().split())
    parent = [i for i in range(v + 1)]
    for i in range(e):
        c, a, b = map(int, sys.stdin.readline().rstrip().split())
        if c:
            if get_parent(parent, a) == get_parent(parent, b):
                print("YES")
            else:
                print("NO")
        else:
            union_parent(parent, a, b)


if __name__ == "__main__":
    solution()
